deepspeed/runtime/zero/stage3.py

```diff
@@ -165,383 +168,328 @@ def _inject_parameters(module, cls):
+        self.trace_complete: bool = False
+        # sequence of submodules/parameters in forward pass + backward pass
+        self.__submodule_order: Iterable[Module] = []
+        self.__param_order: Iterable[__class__.__ParamInTrace] = []
+        self.__most_recent_step_id_param_fetched_for = collections.defaultdict(
+            lambda: int(-1e10))
+        # number of available params, and max number of available params
+        self.__n_available_params: int = 0
+        self.__max_n_available_params: int = max_available_parameters_in_numel
+        # max distance between two use of the module beyond which module is released
+        self.__max_reuse_dist_in_numel: int = max_reuse_distance_in_numel
+        # queue for parameters to fetch. parameters will be popped off the left
+        # side of the dequeue as they are fetched
+        self.__param_queue: Deque[__class__.__ParamInTrace] = None
+        self.__prefetch_bucket_sz: int = prefetch_bucket_sz
+        self.__prefetch_nvme: bool = prefetch_nvme
+        self.hierarchy: int = 0
+
+        # stream that will be used for allgather operations
+        self.__allgather_stream: Stream = allgather_stream
+
+        # limit the number of fetch events that can be queued at once
+        # otherwise, what happens is memory is allocated by the host thread at the
+        # time of the call, but not used until later by the asynchronous cuda stream.
+        # allowing an infinite number of these to queue up causes a lot of memory
+        # pressure that then becomes detrimental to performance.
+        # this is a much less elegant way of fixing this vs something like using
+        # cudaMallocAsync/cudaFreeAsync. Choosing to not expose this to the user now
+        # because ideally in the future its replaced by an async allocation
+        # mechanism which doesnt require any configuration by the user.
+        self.__ongoing_fetch_events: Deque[Event] = collections.deque()
+        self.__max_ongoing_fetch_events: int = 2
+
+    """Tracing and Tracking
+    TODO. consider performing trace before initializing PartitionedParameterCoordinator
+    and passing trace results into constructor. This way all the code in here can
+    just assume that the trace is complete and the results can be entirely
+    immutable.
+
+    Bookkeeping operations used to track where we are in the forward/backward pass
+    """
 
-    def record_trace(self, sub_module):
-        if not self.trace_completed:
-            self.sub_module_trace.append(sub_module.id)
-            self.id_to_sub_module_map[sub_module.id] = sub_module
+    def record_trace(self, sub_module: Module) -> None:
+        """adds sub module to trace"""
+        if self.trace_complete:
+            raise RuntimeError(
+                "attemted to record trace when trace was already complete")
+
+        self.__submodule_order.append(sub_module)
+        for param in sorted(set(iter_params(sub_module)), key=lambda p: p.ds_id):
```

Comment:
Is the sorting here on ds_id important? Does it assume that ds_id represents the order in which the parameters are encountered in forward?

Reply:
i don't think it matters in most cases but its just a cheap way to get additional safety. under the hood, `iter_params` gets the parameters by using `module.named_parameters()` and afaict PyTorch has not explicitly guaranteed any ordering of the returned parameters

to DeepSpeed, we encounter all parameters for a given (leaf) module at once, since we need them all before executing the module. the parameters for a single module are allgathered as a batch. The sorting is to avoid the possibility of parameters for a module being in a different order for different ranks in the batch call. This would cause subtle correctness issues from mixing of data between parameters

Reply:
Also please be aware of needing to know the exact order of params in flattened tensors, since later we need to be able to reconstruct the flattened partitions into a normal model via this map:

https://github.com/microsoft/DeepSpeed/blob/fead387f7837200fefbaba3a7b14709072d8d2cb/deepspeed/runtime/engine.py#L2886-L2926

(it's then used in zero_to_fp32.py to cold extra fp32 weights).

It was already recently updated to not use `module.named_parameters()` but just want to make sure that this function remains in sync with any new sorting. 

Perhaps it's already working just fine, so I'm only bringing it up to make sure it remains so. Thanks.

Reply:
yep that is unaffected. this sorting is for parameter ordering within modules in allgather batches, the ordering within the flattened buffers you describe remains the same